torch/nn/functional.py

```diff
@@ -4124,14 +4124,26 @@ def multi_head_attention_forward(query,                           # type: Tensor
 
     if key_padding_mask is not None:
         attn_output_weights = attn_output_weights.view(bsz, num_heads, tgt_len, src_len)
+        key_padding_mask = key_padding_mask.view(bsz, 1, 1, src_len)
         attn_output_weights = attn_output_weights.masked_fill(
-            key_padding_mask.unsqueeze(1).unsqueeze(2),
+            key_padding_mask,
             float('-inf'),
         )
         attn_output_weights = attn_output_weights.view(bsz * num_heads, tgt_len, src_len)
 
     attn_output_weights = softmax(
         attn_output_weights, dim=-1)
+
+    # mask NaNs with zero
+    if attn_mask is not None and key_padding_mask is not None and attn_mask.dtype == torch.bool:
```

Comment:
I'm still thinking that this fix doesn't exhaust all the cases. For example, it doesn't fix the issue here https://github.com/pytorch/pytorch/issues/24816. The solution I proposed previously works.

Reply:
It's easy to extend it to cover that use case, but it will come at a cost to performance. However, it looks like that issue is the result of erroneous input (empty sequences in the batch) - so my preference would be for it to be handled on the input side as this won't slow things down for other users. I can make the change if needed, although I'd prefer to have a wider disucssion about the change first as it would would introduce a performance regression. 
With regard to your proposed solution, as we discussed with @vadimkantorov earlier, this won't actually work as it doesn't generalize to other data types and also erroneously produces the uniform distribution under some use cases (including in #24816 where it actually produces 0.5, 0.5 weights for the padding). The issue with your solution has also already been discussed twice: https://github.com/pytorch/pytorch/issues/41508#issuecomment-661226577
https://github.com/pytorch/pytorch/pull/42323#discussion_r463858415

Reply:
I have spent some more time on this and I agree that the current implementation will not cover all the valid use cases (eg. time-shifted bucketing could fails as it currently is). I'll do a bit more analysis and update the PR to cover the other situations where masking could produce NaNs.

Reply:
I've handled an additional (valid) edge case, but I still want to give this some more thought.